#' Write fast(a/q) files given Biostrings and ShortRead objects.
#' 
#' @usage write_seq_file(pointer, seqs, seqType, file, compress = FALSE)
#' 
#' @param seqs BioStrings DNAStringSet object generated by trimming from the 
#' pointer object.
#' @param file character Output file name to write sequences.
#' @param compress logical Whether to gzip compress the sequence file when 
#' written or leave it uncompressed. Default is FALSE, or uncompressed.
#' @author Christopher Nobles, Ph.D.

write_seq_files <- function(seqs, file, compress = FALSE){
  packs <- c("Biostrings", "ShortRead")
  packsLoaded <- suppressMessages(sapply(packs, require, character.only = TRUE))
  stopifnot(all(packsLoaded))
  
  seqType <- seq_file_type(file)
  
  if(seqType == "fasta"){
    if(compress){
      if(grepl(".gz$", file)){
        ShortRead::writeFasta(
          seqs, file = file, compress = TRUE, width = max(c(width(seqs), 1)))    
      }else{
        ShortRead::writeFasta(
          seqs, file = paste0(file, ".gz"), 
          compress = TRUE, width = max(c(width(seqs), 1)))
      }
    }else{
      ShortRead::writeFasta(
        seqs, file = file, compress = FALSE, width = max(c(width(seqs), 1)))
    }
  }else{
    if(compress){
      if(grepl(".gz$", file)){
        ShortRead::writeFastq(seqs, file = file, compress = TRUE)
      }else{    
        ShortRead::writeFastq(seqs, file = paste0(file, ".gz"), compress = TRUE)
      }
    }else{
      ShortRead::writeFastq(
        seqs, file = file, compress = FALSE)
    }
  }
}

#' Write fast(a/q) files given zero length files.
#' 
#' @usage write_null_file(file, seqType, writeRandom = FALSE, compress = FALSE)
#'
#' @param file character Output file name to write sequences.
#' @param writeRandom character If FALSE (default), no random output files will 
#' be writen. If character vector, specifies the output file names to write.
#' @param compress logical Whether to gzip compress the sequence file when 
#' written or leave it uncompressed. Default is FALSE, or uncompressed.
#' @author Christopher Nobles, Ph.D.

write_null_file <- function(file, seqType, stat = FALSE, writeRandom = FALSE,
                            compress = FALSE){
  packs <- c("Biostrings", "ShortRead")
  packsLoaded <- suppressMessages(sapply(packs, require, character.only = TRUE))
  stopifnot(all(packsLoaded))
  
  if(seq_file_type(file) == "fasta"){
    null_seqs <- ShortRead::ShortRead()
  }else{
    null_seqs <- ShortRead::ShortReadQ()
  }
  
  # Write empty file
  write_seq_files(null_seqs, file, compress = compress)
  
  # Write zero stat
  if(stat != FALSE){
    sampleName <- unlist(strsplit(file, "/"))
    sampleName <- unlist(
      strsplit(sampleName[length(sampleName)], ".fa", fixed = TRUE))[1]
    write.table(
      data.frame(
        sampleName = sampleName,
        metric = "reads",
        count = length(null_seqs)),
      file = stat,
      sep = ",", row.names = FALSE, col.names = FALSE, quote = FALSE)
  }
  
  # Write empty random sequence file
  if(writeRandom != FALSE){
    null <- lapply(writeRandom, function(x, compress){
      if(seq_file_type(x) == "fasta"){
        rand_null <- ShortRead::ShortRead()
      }else{
        rand_null <- ShortRead::ShortReadQ()
      }
      write_seq_files(rand_null, file = x, compress = compress)
    }, compress = compress)
  }
}

#' Determine output file type from input sequence file name.
#' 
#' seq_file_type(file)
#' 
#' @param file character Vector of file name(s) to get file type.
#' @author Christopher Nobles, Ph.D.

seq_file_type <- function(file){
  file <- unlist(strsplit(file, "/"))
  file <- file[length(file)]
  seqType <- stringr::str_extract(as.character(file), "fa[\\w]*")
  if(any(!seqType %in% c("fa", "fasta", "fastq"))){
    stop("Unrecognized seq file type, please choose '*.fasta' or '*.fastq'.")
  }
  return(ifelse(seqType %in% c("fa", "fasta"), "fasta", "fastq"))
}